{% extends "lesson.html" %}

{% block title %}Funciones Anidadas{% endblock %}

{% block description %}
<p>
Dentro de una funci&oacute;n se puede definir y usar otra funci&oacute;n.  El siguiente
ejemplo muestra una funci&oacute;n auxiliar para simplificar un c&aacute;lculo complicado.</p>

<textarea class="demo">var complicado = function (x) {
    var f = function (y) {
        return y * 3 + 1;
    };
    return f(f(x));
};
var y = complicado(2);
// Ahora y es 22</textarea>

<p>En la definici&oacute;n de una funci&oacute;n se puede referenciar a los argumentos de la
funci&oacute;n padre. Tambi&eacute;n se puede referenciar a las variables <i>fuera</i> de la 
funci&oacute;n anidada.  Un ejemplo pero con la constante
<tt>1</tt> factorizada como una variable.</p>

<textarea class="demo">var complicado = function (x) {
    var c, f;
    c = 1; // constante
    f = function (y) {
        // Aca podemos referenciar a c!
        return y * 3 + c;
    };
    return f(f(x));
};
var y = complicado(2);
// y sigue siendo 22</textarea>

<p>En realidad ese fue el primer ejemplo de una clausura (closure)!  La funci&oacute;n 
externa <i>encierra</i> la funci&oacute;n interna, y sus funciones anidadas
pueden usar variables de la funci&oacute;n externa.</p>

<p>Un caso importante es cuando la funci&oacute;n interna referencia a los argumentos 
desde la funci&oacute;n externa.  Los argumentos son como variables, 
podes referenciarlos desde la funci&oacute;n interna de la misma manera.</p>

<p>El siguiente ejemplo muestra una funci&oacute;n <tt>crear</tt> que recibe un argumento
<tt>x</tt> y retorna una nueva funci&oacute;n.  La funci&oacute;n retornada no recibe
argumentos y retorna <tt>x</tt>.</p>

<textarea class="demo">var crear = function (x) {
    var f = function () {
        return x; // podemos referenciar a x desde aca!
    };
    return f;
};
// crear recibe un argumento, crea una funcion

var g = crear(42);
// g es una funcion que no recibe argumentos ahora

var y = g();
// y es 42</textarea>

<p>En el ejemplo anterior, incluso luego de que la funci&oacute;n <tt>crear</tt> 
termina, la funci&oacute;n que retorna sigue trabajando.  Esa es la 
magia de las clausuras, la referencia a <tt>x</tt> en el alcance externo
sigue siendo v&aacute;lido incluso despu&eacute;s de que la funci&oacute;n externa termina.
Profundo, no?.</p>

<p>Entonces, que es una clausura?  <b>Una clausura es la funci&oacute;n mas el
encierre de su entorno.</b>  En JavaScript cuando estamos enviando valores a una funci&oacute;n, 
le estamos pasando tambi&eacute;n clausuras!</p>

<p class="problem">Definir una funci&oacute;n llamda <tt>genesis</tt> que  
reciba dos argumentos <tt>x</tt> e <tt>y</tt>.  Debe retornar 
una funci&oacute;n que siempre retorne <tt>x + y</tt>.</p>
{% endblock description %}

{% block jsoptions %}
default_code = 'var genesis = function (x, y) {' + nl + '    ' + nl + '};';
test_prelude = '';
add_bool_test('myTypeOf(genesis) === "function"', 'genesis es una funcion');
add_bool_test('myTypeOf(genesis(1, 2)) === "function"', 'genesis(1, 2) es una funcion');
add_assert_test('genesis(1, 2)()', '3');
add_assert_test('genesis(5, 3)()', '8');
{% endblock %}

{#
var genesis = function (x, y) {
    var f = function () {
        return x + y;
    };
    return f;
};
#}

{% block nextlesson %}
var next_lesson = '1070';
{% endblock %}

{% block codeheight %}200{% endblock %}
